MREUnityRuntime/MREUnityRuntimeLib/Assets/AssetCache.cs

```diff
@@ -59,20 +62,41 @@ public GameObject EmptyTemplate()
         /// <inheritdoc cref="GetAsset"/>
         public Object GetAsset(Guid? id)
         {
-            return cache.Find(c => c.Id == id).Asset;
+            return id != null ? cache.Find(c => c.Id == id).Asset : null;
         }
 
         /// <inheritdoc cref="GetId"/>
         public Guid? GetId(Object asset)
         {
-            var id = cache.Find(c => c.Asset == asset).Id;
-            return id != Guid.Empty ? (Guid?)id : null;
+            return asset != null ? cache.Find(c => c.Asset == asset).Id : (Guid?)null;
+        }
+
+        /// <inheritdoc cref="OnCached"/>
+        public void OnCached(Guid id, CacheCallback callback)
+        {
+            var asset = GetAsset(id);
+            if (asset != null)
+            {
+                callback?.Invoke(asset);
+            }
+            else
+            {
+                cacheCallbacks.GetOrCreate(id, () => new List<CacheCallback>(3)).Add(callback);
```

Comment:
`cacheCallbacks` needs to be cleared when app's `FreeResources` method is called.

Reply:
The cache doesn't know which callbacks are for which apps, so the behavior isn't this simple. Resolving the root concern by allowing the callbacks to run, but adding preconditions that return early if the app is destroyed.

---
In reply to: [307024446](https://github.com/microsoft/mixed-reality-extension-unity/pull/92#discussion_r307024446) [](ancestors = 307024446)